sklearn/linear_model/_glm/glm.py

```diff
@@ -20,13 +24,746 @@
+        P = penalty matrix in 1/2 w @ P @ w,
+            for a pure L2 penalty without intercept it equals the identity matrix.
+
+    Note that this solver can naturally deal with sparse X.
+    """
+
+    def setup(self, X, y, sample_weight):
+        """Setup.
+
+        If None, initializes:
+            - self.coef
+        Sets:
+            - self.raw_prediction
+            - self.loss_value
+            - self.g = pointwise gradient
+            - self.h = pointwise hessian
+            - self.gradient
+            - self.sqrt_P = sqrt(l2_reg_strength) * sqrt(P)
+            - self.A_norm
+            - self.X = X
+        """
+        super().setup(X=X, y=y, sample_weight=sample_weight)
+        (self.g, self.h,) = self.linear_loss.base_loss.init_gradient_and_hessian(
+            n_samples=X.shape[0], dtype=X.dtype
+        )
+        # For a general (symmetric) penalty matrix P, we can use any square root of it,
+        # e.g. the Cholesky decomposition. Here, we can use the identity matrix and
+        # handle the intercept.
+        n_features = X.shape[1]
+        # We use a 1-d array instead of a 2-d diagonal array to save memory. We also
+        # simply omit to add a zero at the end for the intercept.
+        self.sqrt_P = np.full(
+            shape=n_features, fill_value=np.sqrt(self.l2_reg_strength), dtype=X.dtype
+        )
+
+        # Initially, set Frobenius norm of A as if h = 1
+        # => ||A||^2 = ||X||^2 + sqrt(l2_reg_strength) ||sqrt(P)||^2
+        # if scipy.sparse.issparse(X):
+        #     A_norm = scipy.sparse.linalg.norm(X) ** 2
+        # else:
+        #     A_norm = scipy.linalg.norm(X) ** 2
+        # A_norm += scipy.linalg.norm(self.sqrt_P) ** 2
+        # self.A_norm = np.sqrt(A_norm)
+        # As this all is a bit costly, we make life easier and start with 1:
+        self.A_norm = 1
+        self.r_norm = 1
+
+        self.X = X  # needed for inner_solve
+
+    def update_gradient_hessian(self, X, y, sample_weight):
+        """Update gradient and hessian.
+
+        Update pointwise gradient and hessian, self.g and self.h,
+        as well as the full gradient, self.gradient.
+        """
+        self.linear_loss.base_loss.gradient_hessian(
+            y_true=y,
+            raw_prediction=self.raw_prediction,  # this was updated in line_search
+            sample_weight=sample_weight,
+            gradient_out=self.g,
+            hessian_out=self.h,
+            n_threads=self.n_threads,
+        )
+        # For non-canonical link functions and far away from the optimum, we take
+        # care that the hessian is positive.
+        eps = np.finfo(y.dtype).eps
+        self.h[self.h <= eps] = eps
+
+        n_features = X.shape[1]
+        # This duplicates a bit of code from LinearModelLoss.gradient.
+        weights, _ = self.linear_loss.weight_intercept(self.coef)
+        self.gradient = np.empty_like(self.coef, dtype=self.coef.dtype)
```

Comment:
we could pre-allocate this buffer in the `setup` step.

Reply:
Actually the call to `self.linear_loss.loss_gradient` in the `line_search` method also allocates its own gradient vector and override this attribute. So as long as we do not have a way to pass a pre-allocated buffer to `self.linear_loss.loss_gradient` there is no point.

The performance impact is probably negligible anyways.